### ISSUE TITLE
Index error in rsolve_poly for solvable recurrence relation

### ISSUE THREAD

**Issue by mmsbrggr** [diofant/diofant#922]
In the following code diofant should provide the solution to a solvable recurrence relation but fails with and index error instead:

from sympy import symbols, Function, sympify, rsolve

```python
from diofant import symbols, Function, sympify, rsolve

n = symbols('n')
x = Function("x")
equation = sympify("-2*n/3 + x(n) - x(n - 1) + 2*(n - 1)**3/3 + 2*(n - 1)**2/3")
solution = rsolve(equation, x(n), init={x(0):0})
print(solution)
```

The terminal output with the error is the following:

Traceback (most recent call last):
  File "rec.py", line 7, in <module>
    solution = rsolve(equation, x(n), init={x(0):0})
  File "/home/marcel/dev/playground/venv/lib/python3.7/site-packages/diofant/solvers/recurr.py", line 709, in rsolve
    return rsolve(f.subs({n: n + abs(k_min)}), y, init)
  File "/home/marcel/dev/playground/venv/lib/python3.7/site-packages/diofant/solvers/recurr.py", line 724, in rsolve
    result = rsolve_hyper(coeffs, i_part, n, symbols=True)
  File "/home/marcel/dev/playground/venv/lib/python3.7/site-packages/diofant/solvers/recurr.py", line 528, in rsolve_hyper
    R = rsolve_ratio(polys, Mul(*denoms), n, symbols=True)
  File "/home/marcel/dev/playground/venv/lib/python3.7/site-packages/diofant/solvers/recurr.py", line 421, in rsolve_ratio
    result = rsolve_poly(numers, f * Mul(*denoms), n, **hints)
  File "/home/marcel/dev/playground/venv/lib/python3.7/site-packages/diofant/solvers/recurr.py", line 304, in rsolve_poly
    solutions = solutions[0]
IndexError: list index out of range

The correct solution to the recurrence equation would be:
`-n**4/6 + n**3/9 + n**2/2+2*n/9`

Changing line 304 in diofant/solvers/recurr.py from `solutions = solutions[0]` to `solutions = solutions[0] if len(solutions) > 0 else {}`, solves the issue.

Python version: 3.7.5
OS: Ubuntu 19.10

**Comment by skirpichev** [comment]
Maybe the problem is the identity ``solve([0], x) == []``.  In this case, the output ``[{}]`` does make sense as well.  Perhaps, it should be used instead, as there _are_ solutions.  Empty list should be reserved for "no solution" case.

**Comment by mmsbrggr** [comment]
That makes more sense to me, yes.
An even smaller example reproducing the error would be:

```python
from diofant import symbols, rsolve_poly
n = symbols('n', integer=True)
rsolve_poly([-9*n-9, 9*n], 9*n*(n - 1)*(n + 1)**2*(9*n + 9), n)
```

**Comment by skirpichev** [comment]
> An even smaller example reproducing ... rsolve_poly

Thank you.  But I would prefer `rsolve()`'s version.  It might be totally unclear, but `rsolve_*()` functions are not intended to be part of the user API (=, roughly, the stuff, that goes from the star-import `from diofant import *`).  These functions are removed from the diofant's namespace in commits of #921.

**Comment by mmsbrggr** [comment]
Good to know. Thank's. Until when can a fix in the production code be expected? I created a fork with a quickfix for this. However, it's of course much cleaner to see this fixed in the production code of diofant.

**Comment by skirpichev** [comment]
This project [adopts the Semantic Versioning numbering scheme](https://diofant.readthedocs.io/en/latest/guide.html#versioning-and-release-procedure), so you can guess from the latest version number (0.10.0) - it's public API should not considered to be stable.  In other words, strictly speaking - there is no production branch(es) and no stable release yet.

The fix will come in 0.11.0 release.  If you wish, I'll do a bugfix release 0.10.1, just to test the workflow for future.  But I dislike to do this just for one issue.  Can you provide some others, you interested in, fixed since the version 0.10.0?  If so, feel free to open an issue for making a bugfix release.
